training_pipeline/models/model.py

```diff
@@ -1,266 +1,287 @@
-    return key + "_xf"
+"""
+    _serving_preprocess, _serving_preprocess_fn, and 
+    _model_exporter functions are defined to provide pre-
+    processing capabilities when the model is served.
+"""
 
 
-def _get_signature(model):
-    signatures = {
-        "serving_default": _get_serve_image_fn(model).get_concrete_function(
-            tf.TensorSpec(
-                shape=[None, 224, 224, 3],
-                dtype=tf.float32,
-                name=_transformed_name(_IMAGE_KEY),
-            )
-        )
+def _serving_preprocess(string_input):
+    decoded_input = tf.io.decode_base64(string_input)
+    decoded = tf.io.decode_jpeg(decoded_input, channels=3)
+    decoded = decoded / 255
+    resized = tf.image.resize(decoded, size=_IMAGE_SHAPE)
+    return resized
+
+
+@tf.function(input_signature=[tf.TensorSpec([None], tf.string)])
+def _serving_preprocess_fn(string_input):
+    decoded_images = tf.map_fn(
+        _serving_preprocess, string_input, dtype=tf.float32, back_prop=False
+    )
+    return {_CONCRETE_INPUT: decoded_images}
+
+
+def _model_exporter(model: tf.keras.Model):
+    m_call = tf.function(model.call).get_concrete_function(
+        tf.TensorSpec(shape=[None, 128, 128, 3], dtype=tf.float32, name=_CONCRETE_INPUT)
+    )
+
+    @tf.function(input_signature=[tf.TensorSpec([None], tf.string)])
+    def serving_fn(string_input):
+        images = _serving_preprocess_fn(string_input)
+        logits = m_call(**images)
+        seg_mask = tf.math.argmax(logits, -1)
+        return {"seg_mask": seg_mask}
+
+    return serving_fn
+
+
+"""
+    _parse_tfr function is defined to parse items from TFRecord
+    files within tf.data pipeline, and _preprocess function will
+    be attatched to the tf.data pipeline to resize input images 
+    before the model
+"""
+
+
+def _parse_tfr(proto):
+    feature_description = {
+        "image": tf.io.FixedLenFeature([], tf.string), 
+        "label": tf.io.FixedLenFeature([], tf.string)
     }
+    rec = tf.io.parse_single_example(proto, feature_description)
+
+    image = tf.io.parse_tensor(rec["image"], tf.float32)
+    label = tf.io.parse_tensor(rec["label"], tf.float32)
 
-    return signatures
+    return {"pixel_values": image, "labels": label}
 
 
-def _get_serve_image_fn(model):
-    @tf.function
-    def serve_image_fn(image_tensor):
-        return model(image_tensor)
+def _preprocess(example_batch):
+    images = example_batch["pixel_values"]
+    images = tf.transpose(
+        images, perm=[0, 1, 2, 3]
+    )  # TF can evaluation the shapes (batch_size,  num_channels, height, width)
+    labels = tf.expand_dims(
+        example_batch["labels"], -1
+    )  # Adds extra dimension, otherwise tf.image.resize won't work.
+    labels = tf.transpose(
+        labels, perm=[0, 1, 2, 3]
+    )  # So, that TF can evaluation the shapes.
 
-    return serve_image_fn
+    images = tf.image.resize(images, _IMAGE_SHAPE)
+    labels = tf.image.resize(labels, _IMAGE_SHAPE)
```

Comment:
As per [this notebook](https://github.com/deep-diver/semantic-segmentation-ml-pipeline/blob/main/notebooks/parse_tfrecords.ipynb), the labels should be of shape `(batch_size, height, width)`.

This means you need to first add a channel dimension to the labels (segmentation maps), then resize it, and then squeeze the channel dimension again. This is what I have done here

https://github.com/deep-diver/semantic-segmentation-ml-pipeline/blob/7a64785b66b144eacbe4a7547e6b310f46def758/tfrecords/create_tfrecords.py#L52

Reply:
this is a **_serving_**preprocess function. Do we need to handle `labels` here too?

Reply:
seems like your explanation is reflected in the code except `transpose`. For instance, the code is doing `expand_dims` => `resize` => `squeeze`.

However, the `resize didn't work without `transpose` by complaining the Tensor doesnt know/have `shape`

Reply:
If this utility is a part of the input data pipeline then this problem shouldn't arise as we verified an input pipeline in the TFRecord parsing notebook.